core/src/main/java/org/elasticsearch/discovery/zen/ZenDiscovery.java

```diff
@@ -932,7 +932,10 @@ public ClusterState execute(ClusterState currentState) {
                     if (modified) {
                         stateBuilder.nodes(nodesBuilder);
                     }
-                    return stateBuilder.build();
+                    currentState = stateBuilder.build();
```

Comment:
actually, I don't want to change this behavior to be honest, I am going to leave it as is. We might to actually do want to send another cluster state as a result